File: IpAddressJava/IpConverter.java
public class IpConverter {
    
    public static int[] decimalToOctets(String decimal) {
        String[] parts = decimal.split("\\.");
        int[] octets = new int[4];
        for (int i = 0; i < 4; i++) {
            octets[i] = Integer.parseInt(parts[i]);
        }
        return octets;
    }
    
    public static int[] binaryToOctets(String binary) {
        int[] octets = new int[4];
        for (int i = 0; i < 4; i++) {
            String octetBinary = binary.substring(i * 8, (i + 1) * 8);
            octets[i] = binaryToDecimal(octetBinary);
        }
        return octets;
    }
    
    public static int[] hexToOctets(String hex) {
        hex = hex.toUpperCase();
        int[] octets = new int[4];
        for (int i = 0; i < 4; i++) {
            String octetHex = hex.substring(i * 2, (i + 1) * 2);
            octets[i] = hexToDecimal(octetHex);
        }
        return octets;
    }
    
    public static String octetsToDecimal(int[] octets) {
        return octets[0] + "." + octets[1] + "." + octets[2] + "." + octets[3];
    }
    
    public static String octetsToBinary(int[] octets) {
        StringBuilder binary = new StringBuilder();
        for (int i = 0; i < 4; i++) {
            if (i > 0) binary.append(".");
            binary.append(decimalToBinary(octets[i], 8));
        }
        return binary.toString();
    }
    
    public static String octetsToHex(int[] octets) {
        StringBuilder hex = new StringBuilder();
        for (int octet : octets) {
            hex.append(decimalToHex(octet, 2));
        }
        return hex.toString();
    }
    
    private static int binaryToDecimal(String binary) {
        int result = 0;
        int power = 0;
        for (int i = binary.length() - 1; i >= 0; i--) {
            if (binary.charAt(i) == '1') {
                result += (1 << power);
            }
            power++;
        }
        return result;
    }
    
    private static int hexToDecimal(String hex) {
        int result = 0;
        int power = 0;
        for (int i = hex.length() - 1; i >= 0; i--) {
            char c = hex.charAt(i);
            int digit;
            if (c >= '0' && c <= '9') {
                digit = c - '0';
            } else {
                digit = c - 'A' + 10;
            }
            result += digit * Math.pow(16, power);
            power++;
        }
        return result;
    }
    
    private static String decimalToBinary(int decimal, int length) {
        StringBuilder binary = new StringBuilder();
        for (int i = length - 1; i >= 0; i--) {
            if ((decimal & (1 << i)) != 0) {
                binary.append("1");
            } else {
                binary.append("0");
            }
        }
        return binary.toString();
    }
    
    private static String decimalToHex(int decimal, int length) {
        String hexChars = "0123456789ABCDEF";
        StringBuilder hex = new StringBuilder();
        for (int i = length - 1; i >= 0; i--) {
            int nibble = (decimal >> (i * 4)) & 0xF;
            hex.append(hexChars.charAt(nibble));
        }
        return hex.toString();
    }
}

File: IpAddressPython/ip_converter.py
def decimal_to_octets(decimal):
    parts = decimal.split('.')
    return [int(part) for part in parts]

def binary_to_octets(binary):
    octets = []
    for i in range(4):
        octet_binary = binary[i * 8:(i + 1) * 8]
        octets.append(binary_to_decimal(octet_binary))
    return octets

def hex_to_octets(hex_str):
    hex_str = hex_str.upper()
    octets = []
    for i in range(4):
        octet_hex = hex_str[i * 2:(i + 1) * 2]
        octets.append(hex_to_decimal(octet_hex))
    return octets

def octets_to_decimal(octets):
    return '.'.join(str(octet) for octet in octets)

def octets_to_binary(octets):
    binary_parts = []
    for octet in octets:
        binary_parts.append(decimal_to_binary(octet, 8))
    return '.'.join(binary_parts)

def octets_to_hex(octets):
    hex_str = ''
    for octet in octets:
        hex_str += decimal_to_hex(octet, 2)
    return hex_str

def binary_to_decimal(binary):
    result = 0
    power = 0
    for i in range(len(binary) - 1, -1, -1):
        if binary[i] == '1':
            result += (1 << power)
        power += 1
    return result

def hex_to_decimal(hex_str):
    result = 0
    power = 0
    for i in range(len(hex_str) - 1, -1, -1):
        c = hex_str[i]
        if c.isdigit():
            digit = int(c)
        else:
            digit = ord(c) - ord('A') + 10
        result += digit * (16 ** power)
        power += 1
    return result

def decimal_to_binary(decimal, length):
    binary = ''
    for i in range(length - 1, -1, -1):
        if (decimal & (1 << i)) != 0:
            binary += '1'
        else:
            binary += '0'
    return binary

def decimal_to_hex(decimal, length):
    hex_chars = '0123456789ABCDEF'
    hex_str = ''
    for i in range(length - 1, -1, -1):
        nibble = (decimal >> (i * 4)) & 0xF
        hex_str += hex_chars[nibble]
    return hex_str
